OpenRA.Mods.Common/Activities/Air/Fly.cs

```diff
@@ -51,35 +50,76 @@ public Fly(Actor self, Target t, WPos? initialTargetPosition = null, Color? targ
 			this.minRange = minRange;
 		}
 
-		public static void FlyToward(Actor self, Aircraft aircraft, int desiredFacing, WDist desiredAltitude, int turnSpeedOverride = -1)
+		public static void FlyTick(Actor self, Aircraft aircraft, int desiredFacing, WDist desiredAltitude, WVec moveOverride, int turnSpeedOverride = -1)
 		{
-			desiredAltitude = new WDist(aircraft.CenterPosition.Z) + desiredAltitude - self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition);
+			var dat = self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition);
+			var move = aircraft.Info.CanHover ? aircraft.FlyStep(desiredFacing) : aircraft.FlyStep(aircraft.Facing);
+			if (moveOverride != WVec.Zero)
+				move = moveOverride;
 
-			var move = aircraft.FlyStep(aircraft.Facing);
-			var altitude = aircraft.CenterPosition.Z;
+			var turnSpeed = turnSpeedOverride > -1 ? turnSpeedOverride : aircraft.TurnSpeed;
+			aircraft.Facing = Util.TickFacing(aircraft.Facing, desiredFacing, turnSpeed);
+
+			if (dat != desiredAltitude)
+			{
+				var maxDelta = move.HorizontalLength * aircraft.Info.MaximumPitch.Tan() / 1024;
+				var deltaZ = (desiredAltitude.Length - dat.Length).Clamp(-maxDelta, maxDelta);
+				move += new WVec(0, 0, deltaZ);
```

Comment:
This still has the issue where it will do the wrong thing (miss the desiredAltitude and possibly move at angles greater than MaximumPitch) if `move.Z != 0`